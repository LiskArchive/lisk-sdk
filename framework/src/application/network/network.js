/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const { getRandomBytes } = require('@liskhq/lisk-cryptography');
const {
	P2P,
	events: {
		EVENT_NETWORK_READY,
		EVENT_NEW_INBOUND_PEER,
		EVENT_CLOSE_INBOUND,
		EVENT_CLOSE_OUTBOUND,
		EVENT_CONNECT_OUTBOUND,
		EVENT_DISCOVERED_PEER,
		EVENT_FAILED_TO_FETCH_PEER_INFO,
		EVENT_FAILED_TO_PUSH_NODE_INFO,
		EVENT_OUTBOUND_SOCKET_ERROR,
		EVENT_INBOUND_SOCKET_ERROR,
		EVENT_UPDATED_PEER_INFO,
		EVENT_FAILED_PEER_INFO_UPDATE,
		EVENT_REQUEST_RECEIVED,
		EVENT_MESSAGE_RECEIVED,
		EVENT_BAN_PEER,
		EVENT_UNBAN_PEER,
	},
} = require('@liskhq/lisk-p2p');
const { lookupPeersIPs } = require('./utils');

const hasNamespaceReg = /:/;

const NETWORK_INFO_KEY_NODE_SECRET = 'network:nodeSecret';
const NETWORK_INFO_KEY_TRIED_PEERS = 'network:triedPeersList';
const DEFAULT_PEER_SAVE_INTERVAL = 10 * 60 * 1000; // 10min in ms

module.exports = class Network {
	constructor({ options, channel, logger, storage }) {
		this.options = options;
		this.channel = channel;
		this.logger = logger;
		this.storage = storage;
		this.secret = null;
	}

	async bootstrap() {
		// Load peers from the database that were tried or connected the last time node was running
		const previousPeersStr = await this.storage.entities.NetworkInfo.getKey(
			NETWORK_INFO_KEY_TRIED_PEERS,
		);
		let previousPeers = [];
		try {
			previousPeers = previousPeersStr ? JSON.parse(previousPeersStr) : [];
		} catch (err) {
			this.logger.error({ err }, 'Failed to parse JSON of previous peers.');
		}

		// Get previous secret if exists
		const secret = await this.storage.entities.NetworkInfo.getKey(
			NETWORK_INFO_KEY_NODE_SECRET,
		);
		if (!secret) {
			this.secret = getRandomBytes(4).readUInt32BE(0);
			await this.storage.entities.NetworkInfo.setKey(
				NETWORK_INFO_KEY_NODE_SECRET,
				this.secret,
			);
		} else {
			this.secret = Number(secret);
		}

		const sanitizeNodeInfo = nodeInfo => ({
			...nodeInfo,
			advertiseAddress: this.options.advertiseAddress,
		});

		const initialNodeInfo = sanitizeNodeInfo(
			await this.channel.invoke('app:getApplicationState'),
		);
		const seedPeers = await lookupPeersIPs(this.options.seedPeers, true);
		const blacklistedIPs = this.options.blacklistedIPs || [];

		const fixedPeers = this.options.fixedPeers
			? this.options.fixedPeers.map(peer => ({
					ipAddress: peer.ip,
					wsPort: peer.wsPort,
			  }))
			: [];

		const whitelistedPeers = this.options.whitelistedPeers
			? this.options.whitelistedPeers.map(peer => ({
					ipAddress: peer.ip,
					wsPort: peer.wsPort,
			  }))
			: [];

		const p2pConfig = {
			nodeInfo: initialNodeInfo,
			hostIp: this.options.hostIp,
			blacklistedIPs,
			fixedPeers,
			whitelistedPeers,
			seedPeers: seedPeers.map(peer => ({
				ipAddress: peer.ip,
				wsPort: peer.wsPort,
			})),
			previousPeers,
			maxOutboundConnections: this.options.maxOutboundConnections,
			maxInboundConnections: this.options.maxInboundConnections,
			peerBanTime: this.options.peerBanTime,
			sendPeerLimit: this.options.sendPeerLimit,
			maxPeerDiscoveryResponseLength: this.options
				.maxPeerDiscoveryResponseLength,
			maxPeerInfoSize: this.options.maxPeerInfoSize,
			wsMaxPayload: this.options.wsMaxPayload,
			secret: this.secret,
		};

		this.p2p = new P2P(p2pConfig);

		this.channel.subscribe('app:state:updated', event => {
			const newNodeInfo = sanitizeNodeInfo(event.data);
			try {
				this.p2p.applyNodeInfo(newNodeInfo);
			} catch (error) {
				this.logger.error(
					`Applying NodeInfo failed because of error: ${error.message ||
						error}`,
				);
			}
		});

		// ---- START: Bind event handlers ----
		this.p2p.on(EVENT_NETWORK_READY, () => {
			this.logger.debug('Node connected to the network');
			this.channel.publish('app:networkReady');
		});

		this.p2p.on(EVENT_CLOSE_OUTBOUND, closePacket => {
			this.logger.debug(
				{
					ipAddress: closePacket.peerInfo.ipAddress,
					wsPort: closePacket.peerInfo.wsPort,
					code: closePacket.code,
					reason: closePacket.reason,
				},
				'EVENT_CLOSE_OUTBOUND: Close outbound peer connection',
			);
		});

		this.p2p.on(EVENT_CLOSE_INBOUND, closePacket => {
			this.logger.debug(
				{
					ipAddress: closePacket.peerInfo.ipAddress,
					wsPort: closePacket.peerInfo.wsPort,
					code: closePacket.code,
					reason: closePacket.reason,
				},
				'EVENT_CLOSE_INBOUND: Close inbound peer connection',
			);
		});

		this.p2p.on(EVENT_CONNECT_OUTBOUND, peerInfo => {
			this.logger.debug(
				{
					ipAddress: peerInfo.ipAddress,
					wsPort: peerInfo.wsPort,
				},
				'EVENT_CONNECT_OUTBOUND: Outbound peer connection',
			);
		});

		this.p2p.on(EVENT_DISCOVERED_PEER, peerInfo => {
			this.logger.trace(
				{
					ipAddress: peerInfo.ipAddress,
					wsPort: peerInfo.wsPort,
				},
				'EVENT_DISCOVERED_PEER: Discovered peer connection',
			);
		});

		this.p2p.on(EVENT_NEW_INBOUND_PEER, peerInfo => {
			this.logger.debug(
				{
					ipAddress: peerInfo.ipAddress,
					wsPort: peerInfo.wsPort,
				},
				'EVENT_NEW_INBOUND_PEER: Inbound peer connection',
			);
		});

		this.p2p.on(EVENT_FAILED_TO_FETCH_PEER_INFO, error => {
			this.logger.error(error.message || error);
		});

		this.p2p.on(EVENT_FAILED_TO_PUSH_NODE_INFO, error => {
			this.logger.trace(error.message || error);
		});

		this.p2p.on(EVENT_OUTBOUND_SOCKET_ERROR, error => {
			this.logger.debug(error.message || error);
		});

		this.p2p.on(EVENT_INBOUND_SOCKET_ERROR, error => {
			this.logger.debug(error.message || error);
		});

		this.p2p.on(EVENT_UPDATED_PEER_INFO, peerInfo => {
			this.logger.trace(
				{
					ipAddress: peerInfo.ipAddress,
					wsPort: peerInfo.wsPort,
				},
				'EVENT_UPDATED_PEER_INFO: Update peer info',
				JSON.stringify(peerInfo),
			);
		});

		this.p2p.on(EVENT_FAILED_PEER_INFO_UPDATE, error => {
			this.logger.error(error.message || error);
		});

		this.p2p.on(EVENT_REQUEST_RECEIVED, async request => {
			this.logger.trace(
				`EVENT_REQUEST_RECEIVED: Received inbound request for procedure ${request.procedure}`,
			);
			// If the request has already been handled internally by the P2P library, we ignore.
			if (request.wasResponseSent) {
				return;
			}
			const hasTargetModule = hasNamespaceReg.test(request.procedure);
			// If the request has no target module, default to app (to support legacy protocol).
			const sanitizedProcedure = hasTargetModule
				? request.procedure
				: `app:${request.procedure}`;
			try {
				const result = await this.channel.invokePublic(sanitizedProcedure, {
					data: request.data,
					peerId: request.peerId,
				});
				this.logger.trace(
					`Peer request fulfilled event: Responded to peer request ${request.procedure}`,
				);
				request.end(result); // Send the response back to the peer.
			} catch (error) {
				this.logger.error(
					`Peer request not fulfilled event: Could not respond to peer request ${
						request.procedure
					} because of error: ${error.message || error}`,
				);
				request.error(error); // Send an error back to the peer.
			}
		});

		this.p2p.on(EVENT_MESSAGE_RECEIVED, async packet => {
			this.logger.trace(
				`EVENT_MESSAGE_RECEIVED: Received inbound message from ${packet.peerId} for event ${packet.event}`,
			);
			this.channel.publish('app:networkEvent', packet);
		});

		this.p2p.on(EVENT_BAN_PEER, peerId => {
			this.logger.error(
				{ peerId },
				'EVENT_MESSAGE_RECEIVED: Peer has been banned temporarily',
			);
		});

		this.p2p.on(EVENT_UNBAN_PEER, peerId => {
			this.logger.error(
				{ peerId },
				'EVENT_MESSAGE_RECEIVED: Peer ban has expired',
			);
		});

		setInterval(async () => {
			const triedPeers = this.p2p.getTriedPeers();
			if (triedPeers.length) {
				await this.storage.entities.NetworkInfo.setKey(
					NETWORK_INFO_KEY_TRIED_PEERS,
					JSON.stringify(triedPeers),
				);
			}
		}, DEFAULT_PEER_SAVE_INTERVAL);

		// ---- END: Bind event handlers ----

		try {
			await this.p2p.start();
		} catch (error) {
			this.logger.fatal(
				{
					message: error.message,
					stack: error.stack,
				},
				'Failed to initialize network',
			);
			process.emit('cleanup', error);
		}
	}

	async request(requestPacket) {
		return this.p2p.request({
			procedure: requestPacket.procedure,
			data: requestPacket.data,
		});
	}

	send(sendPacket) {
		return this.p2p.send({
			event: sendPacket.event,
			data: sendPacket.data,
		});
	}

	async requestFromPeer(requestPacket) {
		return this.p2p.requestFromPeer(
			{
				procedure: requestPacket.procedure,
				data: requestPacket.data,
			},
			requestPacket.peerId,
		);
	}

	sendToPeer(sendPacket) {
		return this.p2p.sendToPeer(
			{
				event: sendPacket.event,
				data: sendPacket.data,
			},
			sendPacket.peerId,
		);
	}

	broadcast(broadcastPacket) {
		return this.p2p.broadcast({
			event: broadcastPacket.event,
			data: broadcastPacket.data,
		});
	}

	getConnectedPeers() {
		return this.p2p.getConnectedPeers();
	}

	getDisconnectedPeers() {
		return this.p2p.getDisconnectedPeers();
	}

	applyPenalty(penaltyPacket) {
		return this.p2p.applyPenalty({
			peerId: penaltyPacket.peerId,
			penalty: penaltyPacket.penalty,
		});
	}

	async stop() {
		// TODO: Unsubscribe 'app:state:updated' from channel.
		this.logger.info('Cleaning network...');

		await this.p2p.stop();
	}
};
