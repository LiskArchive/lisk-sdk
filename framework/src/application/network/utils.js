/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const net = require('net');
const dns = require('dns');

const lookupPromise = (hostname, options) =>
	new Promise((resolve, reject) => {
		dns.lookup(hostname, options, (err, address) => {
			if (err) {
				return reject(err);
			}

			return resolve(address);
		});
	});

const lookupPeersIPs = async (peersList, enabled) => {
	// If peers layer is not enabled there is no need to create the peer's list
	if (!enabled) {
		return [];
	}

	// In case domain names are used, resolve those to IP addresses.
	return Promise.all(
		peersList.map(async peer => {
			if (net.isIPv4(peer.ip)) {
				return peer;
			}

			try {
				const address = await lookupPromise(peer.ip, { family: 4 });
				return {
					...peer,
					ip: address,
				};
			} catch (err) {
				console.error(
					`Failed to resolve peer domain name ${peer.ip} to an IP address`,
				);
				return peer;
			}
		}),
	);
};

module.exports = {
	lookupPeersIPs,
};
