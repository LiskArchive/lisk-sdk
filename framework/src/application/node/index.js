/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

const { config: DefaultConfig } = require('./defaults');
const Chain = require('./chain');
const { migrations } = require('./components');
const BaseModule = require('../../modules/base_module');

/* eslint-disable class-methods-use-this */

class ChainModule extends BaseModule {
	constructor(options) {
		super(options);

		this.chain = null;
	}

	static get alias() {
		return 'chain';
	}

	static get info() {
		return {
			author: 'LiskHQ',
			version: '0.1.0',
			name: 'lisk-framework-chain',
		};
	}

	static get migrations() {
		return migrations;
	}

	static get defaults() {
		return DefaultConfig;
	}

	get events() {
		return [
			'bootstrap',
			'blocks:change',
			'transactions:confirmed:change',
			'signature:change',
			'transactions:change',
			'rounds:change',
			'multisignatures:signature:change',
			'multisignatures:change',
			'delegates:fork',
			'loader:sync',
			'dapps:change',
			'rebuild',
			'processor:sync',
			'processor:deleteBlock',
			'processor:broadcast',
			'processor:newBlock',
		];
	}

	get actions() {
		return {
			calculateSupply: {
				handler: action => this.chain.actions.calculateSupply(action),
			},
			calculateMilestone: {
				handler: action => this.chain.actions.calculateMilestone(action),
			},
			calculateReward: {
				handler: action => this.chain.actions.calculateReward(action),
			},
			getForgerPublicKeysForRound: {
				handler: async action =>
					this.chain.actions.getForgerPublicKeysForRound(action),
			},
			updateForgingStatus: {
				handler: async action => this.chain.actions.updateForgingStatus(action),
			},
			postSignature: {
				handler: async action => this.chain.actions.postSignature(action),
			},
			getForgingStatusForAllDelegates: {
				handler: async () =>
					this.chain.actions.getForgingStatusForAllDelegates(),
			},
			getTransactionsFromPool: {
				handler: async action =>
					this.chain.actions.getTransactionsFromPool(action),
			},
			getTransactions: {
				handler: async action => this.chain.actions.getTransactions(action),
				isPublic: true,
			},
			getSignatures: {
				handler: async () => this.chain.actions.getSignatures(),
				isPublic: true,
			},
			postTransaction: {
				handler: async action => this.chain.actions.postTransaction(action),
			},
			getSlotNumber: {
				handler: async action => this.chain.actions.getSlotNumber(action),
			},
			calcSlotRound: {
				handler: async action => this.chain.actions.calcSlotRound(action),
			},
			getNodeStatus: {
				handler: async () => this.chain.actions.getNodeStatus(),
			},
			getLastBlock: {
				handler: async () => this.chain.actions.getLastBlock(),
				isPublic: true,
			},
			getBlocksFromId: {
				handler: async action => this.chain.actions.getBlocksFromId(action),
				isPublic: true,
			},
			getHighestCommonBlock: {
				handler: async action =>
					this.chain.actions.getHighestCommonBlock(action),
				isPublic: true,
			},
		};
	}

	async load(channel) {
		this.chain = new Chain(channel, this.options);
		await this.chain.bootstrap();
		channel.publish('chain:bootstrap');
	}

	async unload() {
		return this.chain.cleanup();
	}
}

module.exports = ChainModule;
