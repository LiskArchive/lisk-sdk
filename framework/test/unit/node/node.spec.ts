/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { BFT } from '@liskhq/lisk-bft';
import { KVStore } from '@liskhq/lisk-db';
import { TransactionPool } from '@liskhq/lisk-transaction-pool';
import { when } from 'jest-when';
import { TokenModule } from '../../../src/modules';
import { Forger, HighFeeForgingStrategy } from '../../../src/node/forger';
import { Network } from '../../../src/node/network';
import { Node } from '../../../src/node/node';
import { Processor } from '../../../src/node/processor';
import { Synchronizer } from '../../../src/node/synchronizer/synchronizer';
import { genesisBlock } from '../../fixtures';
import * as genesisBlockJSON from '../../fixtures/config/devnet/genesis_block.json';
import { cacheConfig, nodeOptions } from '../../fixtures/node';
import { createMockBus } from '../../utils/channel';

jest.mock('@liskhq/lisk-db');

describe('Node', () => {
	let node: Node;
	let subscribedEvents: any;
	const stubs: any = {};
	const lastBlock = genesisBlock();
	let blockchainDB: KVStore;
	let forgerDB: KVStore;
	let nodeDB: KVStore;

	beforeEach(() => {
		// Arrange
		subscribedEvents = {};

		jest.spyOn(Processor.prototype, 'init').mockResolvedValue(undefined);
		jest.spyOn(Synchronizer.prototype, 'init').mockResolvedValue(undefined);

		blockchainDB = new KVStore('blockchain.db');
		forgerDB = new KVStore('forger.db');
		nodeDB = new KVStore('node.db');

		/* Arranging Stubs start */
		stubs.logger = {
			trace: jest.fn(),
			error: jest.fn(),
			debug: jest.fn(),
			fatal: jest.fn(),
			info: jest.fn(),
			cleanup: jest.fn(),
		};

		stubs.forgerDB = {
			get: jest.fn(),
			put: jest.fn(),
			close: jest.fn(),
		};

		stubs.channel = {
			invoke: jest.fn(),
			subscribe: jest.fn((event, cb) => {
				subscribedEvents[event] = cb;
			}),
			once: jest.fn(),
			registerToBus: jest.fn(),
		};

		when(stubs.channel.invoke)
			.calledWith('app:getComponentConfig', 'cache')
			.mockResolvedValue(cacheConfig as never);

		node = new Node({
			options: nodeOptions,
			genesisBlockJSON,
		});
		node.registerModule(new TokenModule(nodeOptions.genesisConfig));
	});

	describe('constructor', () => {
		it('should throw error when waitThreshold is greater than blockTime', () => {
			const invalidChainOptions = {
				...nodeOptions,
				forging: {
					waitThreshold: 5,
				},
				genesisConfig: {
					...nodeOptions.genesisConfig,
					blockTime: 4,
				},
			};

			expect(
				() =>
					new Node({
						options: invalidChainOptions as any,
						genesisBlockJSON,
					}),
			).toThrow('forging.waitThreshold=5 is greater or equal to genesisConfig.blockTime=4');
		});

		it('should throw error when waitThreshold is same as blockTime', () => {
			const invalidChainOptions = {
				...nodeOptions,
				forging: {
					waitThreshold: 5,
				},
				genesisConfig: {
					...nodeOptions.genesisConfig,
					blockTime: 5,
				},
			};

			expect(
				() =>
					new Node({
						options: invalidChainOptions as any,
						genesisBlockJSON,
					}),
			).toThrow('forging.waitThreshold=5 is greater or equal to genesisConfig.blockTime=5');
		});
	});

	describe('init', () => {
		beforeEach(async () => {
			jest.spyOn(Network.prototype, 'applyNodeInfo');
			jest.spyOn(TransactionPool.prototype, 'start');
			jest.spyOn(node as any, '_startForging');
			// Act
			await node.init({
				bus: createMockBus() as any,
				channel: stubs.channel,
				blockchainDB,
				forgerDB,
				nodeDB,
				logger: stubs.logger,
			});
		});

		it('should initialize scope object with valid structure', () => {
			expect(node).toHaveProperty('_options');
			expect(node).toHaveProperty('_channel');
			expect(node).toHaveProperty('_networkIdentifier');
		});

		describe('_initModules', () => {
			it('should initialize bft module', () => {
				expect(node['_bft']).toBeInstanceOf(BFT);
			});

			it('should initialize forger module', () => {
				expect(node['_forger']).toBeInstanceOf(Forger);
			});

			it('should initialize forger module with high fee strategy', () => {
				expect(node['_forger']['_forgingStrategy']).toBeInstanceOf(HighFeeForgingStrategy);
			});
		});

		it('should invoke Processor.init', () => {
			expect(node['_processor'].init).toHaveBeenCalledTimes(1);
		});

		it('should call "applyNodeInfo" with correct params', () => {
			// Assert
			return expect(node['_networkModule'].applyNodeInfo).toHaveBeenCalledWith({
				height: lastBlock.header.height,
				blockVersion: lastBlock.header.version,
				maxHeightPrevoted: 0,
				lastBlockID: lastBlock.header.id,
			});
		});

		it('should start transaction pool', () => {
			return expect(node['_transactionPool'].start).toHaveBeenCalled();
		});

		it('should start forging', () => {
			return expect(node['_startForging']).toHaveBeenCalled();
		});

		it('should subscribe to "app:network:ready" event', () => {
			return expect(node['_channel'].subscribe).toHaveBeenCalledWith(
				'app:network:ready',
				expect.any(Function),
			);
		});

		it('should subscribe to "app:network:event" event', () => {
			return expect(node['_channel'].subscribe).toHaveBeenCalledWith(
				'app:network:event',
				expect.any(Function),
			);
		});
	});

	describe('getSchema', () => {
		it('should return all schema with currently registered modules', () => {
			const schema = node.getSchema();
			expect(Object.keys(schema.account.properties)).toInclude('token');
			expect(Object.keys(schema.blockHeadersAssets).length).toBeGreaterThanOrEqual(2);
			expect(schema.block).not.toBeUndefined();
			expect(schema.blockHeader).not.toBeUndefined();
			expect(schema.transaction).not.toBeUndefined();
		});
	});

	describe('getRegisteredModules', () => {
		it('should return currently registered modules information', () => {
			const registeredModules = node.getRegisteredModules();
			expect(registeredModules).toHaveLength(1);
			expect(registeredModules[0].name).toEqual('token');
		});
	});

	describe('cleanup', () => {
		beforeEach(async () => {
			// Arrange
			await node.init({
				bus: createMockBus() as any,
				channel: stubs.channel,
				blockchainDB,
				forgerDB,
				nodeDB,
				logger: stubs.logger,
			});
			jest.spyOn(node['_transactionPool'], 'stop');
			jest.spyOn(node['_processor'], 'stop');
			jest.spyOn(node['_synchronizer'], 'stop');
			jest.spyOn(node['_networkModule'], 'cleanup');
		});

		it('should be an async function', () => {
			// Assert
			return expect(node.cleanup.constructor.name).toEqual('AsyncFunction');
		});

		it('should call stop for running tasks', async () => {
			await node.cleanup();
			// Assert
			expect(node['_transactionPool'].stop).toHaveBeenCalled();
			expect(node['_synchronizer'].stop).toHaveBeenCalled();
			expect(node['_processor'].stop).toHaveBeenCalled();
			expect(node['_networkModule'].cleanup).toHaveBeenCalled();
		});
	});

	describe('#_forgingTask', () => {
		beforeEach(async () => {
			await node.init({
				bus: createMockBus() as any,
				channel: stubs.channel,
				blockchainDB,
				forgerDB,
				nodeDB,
				logger: stubs.logger,
			});
			jest.spyOn(node['_forger'], 'delegatesEnabled').mockReturnValue(true);
			jest.spyOn(node['_forger'], 'forge');
			jest.spyOn(node['_synchronizer'], 'isActive', 'get').mockReturnValue(false);
		});

		it('should halt if no delegates are enabled', async () => {
			// Arrange
			(node['_forger'].delegatesEnabled as jest.Mock).mockReturnValue(false);

			// Act
			await node['_forgingTask']();

			// Assert
			expect(stubs.logger.trace).toHaveBeenNthCalledWith(1, 'No delegates are enabled');
			expect(node['_forger'].forge).not.toHaveBeenCalled();
		});

		it('should halt if the client is not ready to forge (is syncing)', async () => {
			// Arrange
			jest.spyOn(node['_synchronizer'], 'isActive', 'get').mockReturnValue(true);

			// Act
			await node['_forgingTask']();

			// Assert
			expect(stubs.logger.debug).toHaveBeenNthCalledWith(1, 'Client not ready to forge');
			expect(node['_forger'].forge).not.toHaveBeenCalled();
		});

		it('should execute forger.forge otherwise', async () => {
			await node['_forgingTask']();

			expect(node['_forger'].forge).toHaveBeenCalled();
		});
	});

	describe('#_startForging', () => {
		beforeEach(async () => {
			await node.init({
				bus: createMockBus() as any,
				channel: stubs.channel,
				blockchainDB,
				forgerDB,
				nodeDB,
				logger: stubs.logger,
			});
			jest.spyOn(node['_forger'], 'loadDelegates');
		});

		it('should load the delegates', async () => {
			await node['_startForging']();
			expect(node['_forger'].loadDelegates).toHaveBeenCalled();
		});

		it('should register a task in Jobs Queue named "nextForge" with a designated interval', async () => {
			await node['_startForging']();

			expect(node['_forgingJob']).not.toBeUndefined();
		});
	});
});
