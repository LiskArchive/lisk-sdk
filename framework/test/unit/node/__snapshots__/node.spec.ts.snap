// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node getEndpoints should not change the exposed endpoints unintentionally 1`] = `
Array [
  "getBlockByID",
  "getBlocksByIDs",
  "getBlockByHeight",
  "getBlocksByHeightBetween",
  "getTransactionByID",
  "getTransactionsByIDs",
  "getLastBlock",
  "getSchema",
  "getRegisteredModules",
  "getNodeInfo",
  "getEvents",
  "proveEvents",
  "postTransaction",
  "getForgingStatus",
  "getTransactionsFromPool",
  "updateForgingStatus",
]
`;
